### ISSUE TITLE
Feature request: extend pillar to use in multi cluster deployment

### ISSUE THREAD

**Issue by porunov** [saltstack-formulas/hadoop-formula#48]
Hello @sroegner , 
I would like to provide PR which will extend an existing pillar options with `clusters` option to deploy several hadoop clusters.

I would like to hear your review and whether you like it or not.

If you do, I can provide the PR. If you don't, please, say why.
 
Here is an example of additional parameters in pillar:

```
hdfs:
  clusters:
    cluster1:
      primary_namenode: 192.168.0.101
      secondary_namenode: 192.168.0.102
      journalnodes:
        - 192.168.0.101
        - 192.168.0.102
        - 192.168.0.103
      zookeeper_connection_string: 192.168.0.101:2181,192.168.0.102:2182,192.168.0.103:2183
      datanodes:
        - 192.168.0.104
        - 192.168.0.105
        - 192.168.0.106
    cluster2:
      primary_namenode: 192.168.1.101
      secondary_namenode: 192.168.1.102
      journalnodes:
        - 192.168.1.101
        - 192.168.1.102
        - 192.168.1.103
      zookeeper_connection_string: 192.168.1.101:2181,192.168.1.102:2182,192.168.1.103:2183
      datanodes:
        - 192.168.1.104
        - 192.168.1.105
        - 192.168.1.106
        
mapred:
  clusters:
    cluster1:
      jobtracker: 192.168.0.101
      jobtracker_on_primary_namenode: Flase
      tasktrackers_on_datanodes: False
      tasktrackers:
        - 192.168.0.104
        - 192.168.0.105
        - 192.168.0.106
    cluster2:
      jobtracker: 192.168.1.101
      jobtracker_on_primary_namenode: True
      tasktrackers_on_datanodes: True
      tasktrackers: 
        - 192.168.1.104
        - 192.168.1.105
        - 192.168.1.106
      
yarn:
  clusters:
    cluster1:
      primary_resourcemanager: 192.168.0.101
      secondary_resourcemanager: 192.168.0.102
      resource_manager_on_namenode: True
      nodemanagers_on_datanodes: True
      zookeeper_connection_string: 192.168.0.101:2181,192.168.0.102:2182,192.168.0.103:2183
      nodemanagers:
        - 192.168.0.104
        - 192.168.0.105
        - 192.168.0.106
    cluster2:
      primary_resource_manager: 192.168.0.101
      secondary_resource_manager: 192.168.0.102
      resource_manager_on_namenode: False
      nodemanagers_on_datanodes: False
      zookeeper_connection_string: 192.168.1.101:2181,192.168.1.102:2182,192.168.1.103:2183
      nodemanagers:
        - 192.168.1.104
        - 192.168.1.105
        - 192.168.1.106
```
        
About mapred:
In case `jobtracker_on_primary_namenode` is equal to True, then `jobtracker` parameter isn't used. Job tracker will be installed on primary namenode.
In case `tasktrackers_on_datanodes` is equal to True then task trackers will be also installed on all data nodes (in addition with `tasktrackers` list).
In case there are no `tasktrackers` parameters in the cluster and `tasktrackers_on_datanodes` is set to False, targeting methods will be used.

About yarn:
In case `resourcemanager_on_namenode` is equal to True, then `primary_resource_manager` and `secondary_resource_manager` parameters are ignored. Resource manager will be installed on a primary namenode or on both primary and secondary namenodes if secondary namenode is specified.
In case `nodemanagers_on_datanodes` is equal to True, then `nodemanagers` will be also installed on all data nodes (in addition with `nodemanagers` list).
In case there are no `nodemanagers` parameter in the cluster and `nodemanagers_on_datanodes` is set to False, targeting methods will be used.

`zookeeper_connection_string` is an optional parameter. In case it is empty Hadoop formula will fetch connection string from Zookeeper formula.

About yarn. If you don't want to complicate it, I can remove yarn high availability feature.

Best regards,
Alexandr

**Comment by sroegner** [comment]
@porunov pls clarify what the final intent is:
- have sth like 6 machines with cluster1 on the first three and cluster2 on the rest
- have sth like 3 machines with cluster1 and cluster2 sharing them
- or both?

**Comment by porunov** [comment]
@sroegner first variant:

> have sth like 6 machines with cluster1 on the first three and cluster2 on the rest

**Comment by porunov** [comment]
Also, I would like to add Hadoop 3 support if you are agree. I.e. support more than 2 NN nodes for a single cluster.
In this way we will not use `primary` and `secondary` namenodes but just `namenodes` list.
In this way we can check a major release to know  whether we can use 1, 2 or many NN nodes.

Here are problems which I see right now with adding Hadoop 3 support:
1. `ha_namenode.sls`. I see that this state designed for convenience, but I am not sure if this state is safe enough. What happens if our primary node crashes and we change it with the new one? In this case our secondary namenode becomes active. Will it be safe to execute this state on the fresh new primary namenode?
Actually, it's not a problem to add Hadoop 3 support for this state (for example. first node will be the primary and other nodes will be secondary). 
2. For backward compatibility we need to support next variables:
`namenode`, `primary_namenode`, `secondary_namenode`, `namenodes`.
I am sure clients will be confused. Personally, I would break backward compatibility in this case. I would support only one parameter `namenodes`. Then I would check a major release and if it is `1` than only the first namenode is used, if it is `2` then 1 or 2 first namenodes are used, if it is `3` then all namenodes are used (for example, or something  similar). Logically the first namenode in the list will be a `primary` (to support `ha_namenode.sls`).
3. Not sure if `ha_namenode_port`, `ha_namenode_http_port`  are needed.

Also, I see several other things which I don't fully understand:
1) Why is `ha_cluster_id` separate from `hdfs` or `hadoop`? Why not to use `hadoop:cluster_id`?
2) Disks lists are looks OK, but still why not to use `hdfs:data_disks` instead of `hdfs_data_disks`.

I understand that the formula is leaving for a long time and it's not nice to break compatibility, but still it's not possible to support compatibilities forever. Something must be changed.
Next must the last alpha release of Hadoop 3 will be produced. Based on their plans stable release will be on 15 August:
https://cwiki.apache.org/confluence/display/HADOOP/Hadoop+3.0.0+release
But the formula doesn't support Hadoop 3 right now.

What do think about all this stuff? Do you agree to let me break compatibility or it's better to still support it?

**Comment by sroegner** [comment]
@porunov 
Just my personal thoughts:
- I don't share the urgency you assign to Hadoop 3 support - even if the release materializes it will not be everybody's first choice right away. But working on it is absolutely doable in a branch which should be brought in via a PR. If that gets too hairy maintaining backwards compatibility (which I do not think is optional) you could also consider forking to a hadoop3 formula altogether.
- Some of the things you mention like the primary ha nn crashing and then reprovisioning I would suggest that Salt might not be the tool of choice for operational aspects where distributed state is hidden away from it inside the application layer. What I am saying is I don't think it is useful to try and manage the full lifecycle of a Hdfs HA cluster in Salt. But maybe I am overly pessimistic.

hth
Steffen

**Comment by porunov** [comment]
@sroegner
I believe that working on one project together is always better than separation. 
I will try to extend the formula by saving backward compatibility.
If it will be too messy then I will consider creating another fork.

Best regards,
Alexandr

**Comment by blbradley** [comment]
I believe you can do this by constructing a better `targeting_method` than the default and assigning some extra grains to your minions that fit your situation.

Salt Mine can collect the IP addresses of minions based on matching for use in states. Using the Mine sounds much easier than editing the pillar on IP address change.

**Comment by porunov** [comment]
@blbradley , It depends. For me it's more convenient to use pillar because all info are collected in one place which is easy to edit. I will extend pillar options for now. Maybe in the future we can extend targeting to be used in multi cluster deployment.

**Comment by blbradley** [comment]
@porunov I have set up separate HDFS clusters for development and production using what I described. Targeting does not need to be extended.

**Comment by porunov** [comment]
Resolved by #51

**Comment by blbradley** [comment]
Yay
